File: bindings/ocaml/keystone.ml
open Ctypes
open Foreign

module B = Ffi_bindings.Bindings(Ffi_generated)
module Types = B.T

type encoded_result =
  {
    encoding: int array;
    encoding_size : int;
    stat_count : int;
  }

let ks_arch_supported arch =
  B.ks_arch_supported_ arch

let ks_version major minor =
  let major = Unsigned.UInt.of_int major in
  let minor = Unsigned.UInt.of_int minor in
  let u = B.ks_version_ (allocate uint major) (allocate uint minor) in
  Unsigned.UInt.to_int u

(* TODO: better handling of error value *)
let ks_option engine opttype optvalue =
  match opttype with
  | Types.KS_OPT_SYNTAX -> B.ks_option_ engine Types.KS_OPT_SYNTAX optvalue

let ks_strerror err =
  B.ks_strerror_ err

let ks_open arch ?(endian=Types.KS_MODE_LITTLE_ENDIAN) mode =
  let check_endian =
    match endian with
    | Types.KS_MODE_BIG_ENDIAN ->
       begin
         let m = Ffi_generated_types.constant (Types.string_of_ks_mode Types.KS_MODE_BIG_ENDIAN) int64_t in
         let m' = Ffi_generated_types.constant (Types.string_of_ks_mode mode) int64_t in
         Result.Ok(Int64.add m m')

       end
    | Types.KS_MODE_LITTLE_ENDIAN -> Result.Ok(Ffi_generated_types.constant (Types.string_of_ks_mode mode) int64_t)
    | _ -> Result.Error("Non-endian mode passed to endian arg of ks_open: use KS_MODE_BIG_ENDIAN or KS_MODE_LITTLE_ENDIAN")
  in
  let mode =
    match check_endian with
    | Result.Ok a -> a
    | Result.Error e -> failwith e
  in
  let engine = allocate_n ~count:1 ((ptr Types.ks_engine)) in

  match (B.ks_open_ arch mode engine) with
  | Types.KS_ERR_OK -> Result.Ok(!@ engine)
  | _ as err -> Result.Error(ks_strerror err)


let ks_close engine = B.ks_close_ engine

let ks_errno engine = B.ks_err_ engine


let ks_asm engine str addr =
  let addr' = Int64.of_int addr in
  let encoding = allocate_n ~count:1 (ptr uchar) in
  let encoding_size = allocate size_t (Unsigned.Size_t.of_int 0) in
  let stat_count = allocate size_t (Unsigned.Size_t.of_int 0) in
  match (B.ks_asm_ engine str addr' encoding encoding_size stat_count) with
  | 0 -> begin
      let iencoding_size = Unsigned.Size_t.to_int (!@ encoding_size) in
      let istat_count = Unsigned.Size_t.to_int (!@ stat_count) in
      let encodedasm =
        CArray.from_ptr (!@ encoding) (Unsigned.Size_t.to_int (!@ encoding_size))
        |> CArray.to_list |> Array.of_list  (* No map in CArray *)
        |> Array.map (fun c -> Unsigned.UChar.to_int c)
      in
      B.ks_free_ (to_voidp (!@ encoding));
      Result.Ok({encoding = encodedasm; encoding_size =  iencoding_size; stat_count = istat_count})
    end
  | _ -> let err = ks_errno engine |> ks_strerror  in
         Result.Error err

let asm_array_to_string a =
  Array.fold_left (fun str c -> let t = Printf.sprintf "%02x " c in str^t) "" a

File: bindings/haskell/src/Keystone.hs
{-|
Module      : Keystone
Description : The Keystone assembler engine.
Copyright   : (c) Adrian Herrera, 2016
License     :  GPL-2

Keystone is a lightweight multi-platform, multi-architecture assembler
framework.

Further information is available at <http://www.keystone-engine.org>.
-}
module Keystone
    ( -- * Assembler control
      Assembler
    , Engine
    , Architecture(..)
    , Mode(..)
    , OptionType(..)
    , OptionValue(..)
    , runAssembler
    , open
    , option
    , assemble

      -- * Error handling
    , Error(..)
    , errno
    , strerror

      -- * Misc.
    , version
    ) where

import Control.Monad.Trans.Class (lift)
import Control.Monad.Trans.Except (runExceptT, throwE)
import Data.ByteString (ByteString, packCStringLen)
import Data.List (intercalate)
import Foreign

import Keystone.Internal.Core
import Keystone.Internal.Keystone

-------------------------------------------------------------------------------
-- Assembler control
-------------------------------------------------------------------------------

-- | Run the Keystone assembler and return a result on success, or an 'Error'
-- on failure.
runAssembler :: Assembler a         -- ^ The assembler code to execute
             -> IO (Either Error a) -- ^ A result on success, or an 'Error' on
                                    -- failure
runAssembler =
    runExceptT

-- | Create a new instance of the Keystone assembler.
open :: Architecture        -- ^ CPU architecture
     -> [Mode]              -- ^ CPU hardware mode
     -> Assembler Engine    -- ^ A 'Keystone' engine on success, or an 'Error'
                            -- on failure
open arch mode = do
    (err, ksPtr) <- lift $ ksOpen arch mode
    if err == ErrOk then
        -- Return a pointer to the Keystone engine if ksOpen completed
        -- successfully
        lift $ mkEngine ksPtr
    else
        -- Otherwise return an error
        throwE err

option :: Engine        -- ^ 'Keystone' engine handle
       -> OptionType    -- ^ Type of option to set
       -> OptionValue   -- ^ Option value corresponding with the type
       -> Assembler ()  -- ^ An 'Error' on failure
option ks optType optValue = do
    err <- lift $ ksOption ks optType optValue
    if err == ErrOk then
        return ()
    else
        throwE err

-- | Assemble a list of statements.
assemble :: Engine                      -- ^ 'Keystone' engine handle
         -> [String]                    -- ^ List of statements to assemble.
         -> Maybe Word64                -- ^ Optional address of the first
                                        -- assembly instruction
         -> Assembler (ByteString, Int) -- ^ Returns the encoded input assembly
                                        -- string and the number of statements
                                        -- successfully processed. Returns an
                                        -- 'Error' on failure
assemble ks stmts addr = do
    let string = intercalate ";" stmts
    (res, encPtr, encSize, statCount) <- lift $ ksAsm ks string (maybeZ addr)
    if res == 0 then do
        -- If ksAsm completed successfully, pack the encoded bytes into a
        -- ByteString. Once the encoded bytes have been packed the original
        -- encoded bytes can be freed. The ByteString is returned with the
        -- statement count
        bs <- lift $ packCStringLen (castPtr encPtr, encSize)
        lift $ ksFree encPtr
        return (bs, statCount)
    else do
        -- On failure, call errno for error code
        err <- errno ks
        throwE err
    where maybeZ = maybe 0 id

-------------------------------------------------------------------------------
-- Misc.
-------------------------------------------------------------------------------

-- | Combined API version & major and minor version numbers. Returns a
-- hexadecimal number as (major << 8 | minor), which encodes both major and
-- minor versions.
version :: Int
version =
    ksVersion nullPtr nullPtr

-- | Report the 'Error' number when some API function failed.
errno :: Engine             -- ^ 'Keystone' engine handle
      -> Assembler Error    -- ^ The last 'Error' code
errno =
    lift . ksErrno

-- | Return a string describing the given 'Error'.
strerror :: Error   -- ^ The 'Error' code
         -> String  -- ^ Description of the error code
strerror =
    ksStrerror
